backend/de.metas.monitoring/src/main/java/de/metas/monitoring/adapter/MicrometerPerformanceMonitoringService.java

```diff
@@ -42,58 +44,92 @@
 public class MicrometerPerformanceMonitoringService implements PerformanceMonitoringService
 {
 	private final MeterRegistry meterRegistry;
-	private final Optional<APMPerformanceMonitoringService> apmPerformanceMonitoringService;
+	private final ThreadLocal<Integer> depth = ThreadLocal.withInitial(() -> 0);
+	private final ThreadLocal<String> initiator = ThreadLocal.withInitial(() -> "");
+	private final ThreadLocal<String> initiatorWindow = ThreadLocal.withInitial(() -> "");
+	private final ThreadLocal<ArrayList<String>> calledBy = ThreadLocal.withInitial(() -> new ArrayList<>());
+	private final ThreadLocal<Boolean> isInitiatorLabelActive = ThreadLocal.withInitial(() -> false);
```

Comment:
considering introducing an class which holds depth, initiator etc data.
The class shall have proper methods for adding, removing things which reveal ur intentions.
Here you could have only one ThreadLocal of that class.